### PR TITLE
Improve TECS Height and Speed Accuracy

### PR DESCRIPTION
**Pull request by priseborough**
This PR implements the TECS control loop changes from https://github.com/ArduPilot/ardupilot/pull/15441 and omits the acceleration control method implementation. A separate PR will be created with the acceleration vector changes when this PR is merged.

See  https://github.com/ArduPilot/ardupilot/pull/15441 for details.

Up and away performance of the new TECS control loops has already had extensive testing. Flight testing of the landing flare is required.

This is the old pitch demand calculation control loop architecture:

[TECS pitch demand - current.pdf](https://github.com/ArduPilot/ardupilot/files/9996219/TECS.pitch.demand.-.current.pdf)

This is the new pitch demand calculation control loop architecture:

[TECS pitch demand - proposed.pdf](https://github.com/ArduPilot/ardupilot/files/9996210/TECS.pitch.demand.-.proposed.pdf)

### CODE REVIEW

File: libraries/AP_TECS/AP_TECS.cpp [hunk 1]
```diff
@@ -264,6 +264,33 @@ const AP_Param::GroupInfo AP_TECS::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("PTCH_FF_K", 30, AP_TECS, _pitch_ff_k, 0.0),
 
+    // @Param: LAND_PTRIM
+    // @DisplayName: Pitch angle for level flight in landing confiuration
```
Comment (tridge) [inline libraries/AP_TECS/AP_TECS.cpp:268]: spelling of configuration

File: libraries/AP_TECS/AP_TECS.cpp [hunk 2]
```diff
@@ -56,7 +56,7 @@ const AP_Param::GroupInfo AP_TECS::var_info[] = {
     // @Range: 0.0 0.5
     // @Increment: 0.02
     // @User: Advanced
-    AP_GROUPINFO("INTEG_GAIN", 4, AP_TECS, _integGain, 0.1f),
+    AP_GROUPINFO("INTEG_GAIN", 4, AP_TECS, _integGain, 0.3f),
```
Comment (tridge) [inline libraries/AP_TECS/AP_TECS.cpp:59]: do you think we need to do any param conversion from existing params? Note that hardly anybody tunes TECS ....

**Review by tridge — CHANGES_REQUESTED** [review]
first tests in RF look good! I will fly this on my Yak55 during the week

File: libraries/AP_Landing/AP_Landing.cpp [hunk 3]
```diff
@@ -145,6 +145,15 @@ const AP_Param::GroupInfo AP_Landing::var_info[] = {
...
     AP_GROUPINFO("OPTIONS", 16, AP_Landing, _options, 0),
 
+    // @Param: FLARE_AIM
+    // @DisplayName: Flare aim point adjustment percentage.
+    // @Description: This parameter controls how much the aim point is moved to allow for the time spent in the flare manoeuvre. When set to 100% the aim point is adjusted on the assumption that the flare sink rate controller instantly achieves the sink rate set by TECS_LAND_SINK. when set to 0%, no aim point adjustment is made. If the plane consistently touches down short of the aim point reduce the parameter and vice verse.
+    // @Range: 0 100
+    // @Units: %
+    // @Increment: 1
+    // @User: Advanced
+    AP_GROUPINFO("FLARE_AIM", 17, AP_Landing, flare_effectivness_pct, 50),
```
Comment (IamPete1) [inline libraries/AP_Landing/AP_Landing.cpp:155]: This is somehow a factor of TECS response time I think? I wonder if we could calculate a value.

File: libraries/AP_TECS/AP_TECS.cpp [hunk 4]
```diff
@@ -264,6 +264,33 @@ const AP_Param::GroupInfo AP_TECS::var_info[] = {
...
+    // @Range: -10 15
+    // @Units: deg
+    // @Increment: 1
+    // @User: Advanced
+    AP_GROUPINFO("FLARE_HGT", 32, AP_TECS, _flare_holdoff_hgt, 1.0f),
+
+    // @Param: HDEM_TCONST
+    // @DisplayName: Height Demand Time Constant
+    // @Description: This sets the time constant of the low pass filter that is applied to the height demand input when bit 1 of TECS_OPTIONS is not selected.
+    // @Range: -1.0 5.0
```
Comment (IamPete1) [inline libraries/AP_TECS/AP_TECS.cpp:288]: negative time constant?

File: libraries/AP_TECS/AP_TECS.cpp [hunk 5]
```diff
@@ -245,8 +245,8 @@ const AP_Param::GroupInfo AP_TECS::var_info[] = {
 
     // @Param: OPTIONS
     // @DisplayName: Extra TECS options
-    // @Description: This allows the enabling of special features in the speed/height controller
-    // @Bitmask: 0:GliderOnly
+    // @Description: This allows the enabling of special features in the speed/height controller.
+    // @Bitmask: 0:GliderOnly,1:No Height Demand Smoothing
```
Comment (IamPete1) [inline libraries/AP_TECS/AP_TECS.cpp:249]: I don't think this option made it?

File: libraries/AP_TECS/AP_TECS.cpp (hunk 2 above)
Comment (priseborough) [reply libraries/AP_TECS/AP_TECS.cpp:59]: Probably not. The new TECs can use a larger I value without running into limit cycle issues, hence the change to the default value. If the user is already using a larger I value, then we don't need to change their parameter value.

File: libraries/AP_Landing/AP_Landing.cpp (hunk 3 above)
Comment (priseborough) [reply libraries/AP_Landing/AP_Landing.cpp:155]: Unfortunately  its a function of aircraft lift curve slope, approach speed, glide slope, pitch loop tuning, TECS tuning, etc. The reason I added this parameter is that after the aircraft is tuned and a standardised glide slope and airspeed is determined, this parameter can then be used to fine tune the landing accuracy,

File: libraries/AP_TECS/AP_TECS.cpp (hunk 4 above)
Comment (priseborough) [reply libraries/AP_TECS/AP_TECS.cpp]: Whoops. I removed the - sign and also added some code to protect against someone entering a too small value that would destabilise the LPF that is applied to the height demand.

File: libraries/AP_TECS/AP_TECS.cpp (hunk 5 above)
Comment (priseborough) [reply libraries/AP_TECS/AP_TECS.cpp]: Fixed

**Review by tridge — APPROVED** [review]

File: libraries/AP_Landing/AP_Landing.cpp [hunk 6]
```diff
@@ -145,6 +145,15 @@ const AP_Param::GroupInfo AP_Landing::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("OPTIONS", 16, AP_Landing, _options, 0),
 
+    // @Param: FLARE_AIM
+    // @DisplayName: Flare aim point adjustment percentage.
+    // @Description: This parameter controls how much the aim point is moved to allow for the time spent in the flare manoeuvre. When set to 100% the aim point is adjusted on the assumption that the flare sink rate controller instantly achieves the sink rate set by TECS_LAND_SINK. when set to 0%, no aim point adjustment is made. If the plane consistently touches down short of the aim point reduce the parameter and vice verse.
```
Comment (magicrub) [inline libraries/AP_Landing/AP_Landing.cpp:150]: spelling

**Comment by tridge** [comment]
I tested this on a Yak55 and the improvement in landing performance is huge. Really nice!

**Comment by magicrub** [comment]
[Applause]

**Comment by priseborough** [comment]
@sprivera Not applying the LAND_PITCH_CD during approach is deliberate design decision. It should only be applied during flare as per the parameter description. If you are not getting a proper flare, then this indicates an issue with either pitch controller or TECS tuning that was previously hidden by the way the old TECS was bringing the nose up during the landing approach. Please supply a log.

The  TECS_LAND_PTRIM parameter was included in this PR by mistake. It was supposed to be included in a yet to be submitted PR focussing on improving landing accuracy for aircraft using flaps during approach. I will remove it. Thanks for pointing it out.

**Comment by sprivera** [comment]
I fixed my landing issue, my problem was similar to this one https://github.com/ArduPilot/ardupilot/pull/20792#issuecomment-1270982461 . 

I'm not sure if this may cause problems, but I found that when repeating a landing, the flare counter is not reset and therefore, during flare the demanded altitude is not right. I've attached the log file.
image
[2022-11-23 11-31-49.zip](https://github.com/ArduPilot/ardupilot/files/10077832/2022-11-23.11-31-49.zip)

I've fixed this by adding _flare_counter = 0 after this line. https://github.com/ArduPilot/ardupilot/blob/b12cd48843ec8d95dfd31ddda9eea59f9cdc1932/libraries/AP_TECS/AP_TECS.cpp#L572

That said, thanks for the improvements in the code, the tracking performance is much better.

**Comment by IamPete1** [comment]
@sprivera Looks like we should replace `_flare_counter` with a bool. There is a wrapping issue using uint8_t. Could you do a PR?

**Comment by priseborough** [comment]
@sprivera and @IamPete1 I've submitted a fix https://github.com/ArduPilot/ardupilot/pull/22299

**Pull request merged**